crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/next_target.rs

```diff
@@ -0,0 +1,725 @@
+        let mut rng = thread_rng();
+
+        // - Create N remote nodes.
+        let mut remote_nodes: Vec<_> = (0..n)
+            .map(|_| {
+                let url = random_url(&mut rng);
+                let url = TxUrl::from(url.as_str());
+                let purl = kitsune_p2p_proxy::ProxyUrl::from_full(url.as_str()).unwrap();
+                let target = GossipTgt::new(vec![], Tx2Cert::from(purl.digest()));
+                Node { target, url }
+            })
+            .collect();
+
+        let mut metrics = Metrics::new();
+
+        // - Pop the last node off the list.
+        let last = remote_nodes.pop().unwrap();
+
+        // - Record successful initiate rounds for the rest of the nodes.
+        for node in remote_nodes.iter() {
+            let cert = node.cert().clone();
+            metrics.record_initiate(cert.clone());
+            metrics.record_success(cert);
+        }
+
+        // - Push the last node back into the remote nodes.
+        remote_nodes.push(last);
+
+        let r = next_remote_node(remote_nodes.clone(), &metrics, tuning_params_no_delay());
+
+        // - Expect the last node to be chosen because it was never gossiped with.
+        assert_eq!(r, remote_nodes.last().cloned());
+    }
+
+    #[test]
+    /// Test we break ties between never talked
+    /// to nodes by randomly choosing one.
+    fn randomly_break_ties() {
+        let mut rng = thread_rng();
+
+        // - Create 100 remote nodes.
+        let mut remote_nodes: Vec<_> = (0..100)
+            .map(|_| {
+                let url = random_url(&mut rng);
+                let url = TxUrl::from(url.as_str());
+                let purl = kitsune_p2p_proxy::ProxyUrl::from_full(url.as_str()).unwrap();
+                let target = GossipTgt::new(vec![], Tx2Cert::from(purl.digest()));
+                Node { target, url }
+            })
+            .collect();
+
+        let mut metrics = Metrics::new();
+
+        // - Pop the last two nodes off the list.
+        let last = remote_nodes.pop().unwrap();
+        let second_last = remote_nodes.pop().unwrap();
+
+        // - Record successful initiate rounds for the rest of the nodes.
+        for node in remote_nodes.iter() {
+            let cert = node.cert().clone();
+            metrics.record_initiate(cert.clone());
+            metrics.record_success(cert);
+        }
+
+        // - Push the last two nodes back into the remote nodes.
+        remote_nodes.push(second_last.clone());
+        remote_nodes.push(last);
+
+        // - Check we don't always get the second last node.
+        // The second last is returned always when not randomly breaking ties.
+        let mut always_the_last = true;
+        for _ in 0..100 {
+            let r =
+                next_remote_node(remote_nodes.clone(), &metrics, tuning_params_no_delay()).unwrap();
+            if r != second_last {
+                always_the_last = false;
+            }
+        }
```

Comment:
Seems this would be more robust, it seems possible that we could have a regression where the last node is always picked, instead of the second to last node.

```suggestion
        let mut chose_last = false;
        let mut chose_second_last = false;
        for _ in 0..100 {
            let r =
                next_remote_node(remote_nodes.clone(), &metrics, tuning_params_no_delay()).unwrap();
            if r == last {
                chose_last = true;
            } else if r == second_last {
                chose_second_last = true;
            }
        }
```

Reply:
Oh yeah and then I forgot to add the check below that both of these are true, but you get the idea.